app/src/main/java/ch/sheremet/dasfoo/rover/android/client/MainActivity.java

```diff
@@ -5,64 +5,78 @@
 import android.os.Bundle;
 import android.support.v7.app.AppCompatActivity;
 import android.text.TextUtils;
+import android.util.Log;
 import android.view.View;
 import android.view.inputmethod.InputMethodManager;
 import android.widget.Button;
 import android.widget.EditText;
 import android.widget.TextView;
 import android.widget.Toast;
 
-import java.util.concurrent.TimeUnit;
+import com.google.android.gms.security.ProviderInstaller;
 
+import ch.sheremet.dasfoo.rover.android.client.grpc.authentication.GrpcProviderInstaller;
 import ch.sheremet.dasfoo.rover.android.client.grpc.task.AbstractGrpcTaskExecutor;
 import ch.sheremet.dasfoo.rover.android.client.grpc.task.EncodersReadingTask;
 import ch.sheremet.dasfoo.rover.android.client.grpc.task.GettingBoardInfoTask;
-import ch.sheremet.dasfoo.rover.android.client.grpc.task.IOnGrpcTaskCompleted;
+import ch.sheremet.dasfoo.rover.android.client.grpc.task.GrpcConnection;
 import ch.sheremet.dasfoo.rover.android.client.grpc.task.MovingRoverTask;
-import dasfoo.grpc.roverserver.nano.RoverServiceGrpc;
-import io.grpc.ManagedChannel;
-import io.grpc.ManagedChannelBuilder;
 
-public class MainActivity extends AppCompatActivity implements IOnGrpcTaskCompleted {
+public class MainActivity extends AppCompatActivity {
```

Comment:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [This class has too many methods, consider refactoring it.](https://www.codacy.com/app/dot-doom/rover-android-client/file/2661510811/issues/source?bid=3324652&fileBranchId=3330575#l25)